### DISCUSSION TITLE
Get list of highlighted things from mini.highpattern

### DISCUSSION THREAD

**Discussion by beaumccartney** [nvim-mini/mini.nvim#518] [Q&A] [answered]
One think I really like about folke's todo-comments is the ability to put todos in a quickfix list or telescope finder. I ended up writing my own mini.pick finder that lets me choose todos from todo-comments.

Would it be possible to get a list of matched items for a given set of highlighters from mini.hipatterns? Once the list of matches (plus location) is in hand its easy enough to pipe them to mini.pick or a qf list or whatever.

**Comment by echasnovski** [comment]
Yeah, this is possible, but on a more low level. All highlighting is done via extmarks (see `:h extmarks`) in a single namespace (see `:h namespace`). Right now it has name "MiniHipatternsHighlight" by which you can get namespace id via `vim.api.nvim_get_namespaces().MiniHipatternsHighlight`. After that, usage built-in tools to get extmarks data should be doable (like calling `:h nvim_buf_get_extmarks()` for every present buffer).

I think the best approach here would be to fix and "export" namespace name to allow this kind of manipulations.

**Comment by echasnovski** [comment, marked as answer]
@BeauSLM, thanks again for the suggestion! It prompted me thinking about the best way to do this, and after couple of refactor rounds this is now possible with [`get_matches()`](https://github.com/echasnovski/mini.nvim/blob/7af4878bd0b19fcef385e45715f418165eff1218/doc/mini-hipatterns.txt#L389).

As a heads up, here is how it can be used in a picker (which will be a part of an upcoming 'mini.extra' module):

![screenshot_2023-10-24_15:36:07](https://github.com/echasnovski/mini.nvim/assets/24854248/6f5db8d0-6cc7-4272-979a-85e8aa8f419f)

**Reply by beaumccartney** [reply]
Looks amazing! Glad my idea was helpful.

**Reply by beaumccartney** [reply]
so I may work with it until the module comes out, what's the source for the picker? I've been fiddling for a bit and I'm not sure how you got the highlights on the left

**Reply by echasnovski** [reply]
Highlights are created inside custom `source.show()` with extmarks using items' `hl_group` field.

Here is the [full source](https://github.com/echasnovski/nvim/blob/695583ce6b5e918813423663fd04f69b29a605f0/lua/mini-dev/extra.lua#L904-L949). Beware that the exported functionality and names might change.

**Comment by beaumccartney** [comment]
One usecase I imagined for a feature like this is "hunting" for todos in the project, e.g. trying to fix all todos before a release or something like that. However this would require a project-wide search (backed by ripgrep or something), which seems like its out of the scope of what `hipatterns` does.

I'm honestly not sure how something like this could be supported, thoughts?

**Reply by beaumccartney** [reply]
One workflow could be go file-by-file, but then how to figure out what files have todo-type patterns in them?

`hipatterns` provides a nice way to define what a TODO looks like (or generally a pattern of interest), is it a good idea to decouple that from the actual highlighting?

This kind of intersects with project-wide search and maybe replace, but with a nice way to define search patterns and give meaning to what they represent.

**Reply by echasnovski** [reply]
Yes, the fact that it doesn't scan all files but only enabled buffers is how 'mini.hipatterns' works (and what initially was the motivation to not implement this). And enabling search in some other buffers is indeed out of scope.

To hunt for TODOs in a project, I'd use a good old grepping: either with `MiniPick.builtin.grep()` or `MiniPick.builtin.grep_live()`.
